mathics/eval/tensors.py

```diff
@@ -76,6 +74,85 @@ def get_dimensions(expr, head=None):
         return [len(expr.elements)] + sub
 
 
+def to_std_sparse_array(sparse_array, evaluation: Evaluation):
+    if sparse_array.elements[2] == Integer0:
+        return sparse_array
+    else:
+        return Expression(
+            SymbolSparseArray, Expression(SymbolNormal, sparse_array)
+        ).evaluate(evaluation)
+
+
+def unpack_outer(item, rest_lists, current, level: int, const_etc: tuple):
+    """
+    Recursively unpacks lists to evaluate outer product.
+    ------------------------------------
+
+    Unlike direct products, outer (tensor) products require traversing the
+    lowest level of each list, hence we recursively unpacking lists until
+    the lowest level is reached.
+
+    Parameters:
+
+    ``item``: the current item to be unpacked (if not at lowest level), or joined
+    to current (if at lowest level)
+
+    ``rest_lists``: the rest of lists to be unpacked
+
+    ``current``: the current lowest level elements
+
+    ``level``: the current level (unused yet, will be used in
+    ``Outer[f_, lists__, n_]`` in the future)
+
+    ``const_etc``: a tuple of functions used in unpacking, remains constant

```

Comment:
One thing that is done for this kind of thing, is to have an inner function that is recursive and the other function then just serves to make the these variables accessible within the inner recursive function. 

```python
def unpack_outer(item, rest_lists, current, level: int, const_etc: tuple) -> BaseElement:
    (
        cond_next_list,  # return True when the next list should be unpacked
        get_elements,  # get elements of list, tuple, ListExpression, etc.
        apply_head,  # e.g. lambda elements: Expression(head, *elements)
        apply_f,  # e.g. lambda current: Expression(f, *current)
        join_elem,  # join current lowest level elements (i.e. current) with a new one
        if_nested,  # True for result as nested list ({{a,b},{c,d}}), False for result as flattened list ({a,b,c,d}})
        evaluation,  # evaluation: Evaluation
    ) = const_etc

     def _unpack_outer(item, reset_lists, current, level: int) -> BaseElement:
             # use con_next_list, get_elements, apply_head and so on.
	     # For recursive calls, use _unpack_outer()


     _unpack_outer(item, rest_lists, current, leven)